src/core/dynamic_system.jl

```diff
@@ -1,60 +1,95 @@
 
 **Fields**
-- `H` -- the Hamiltonian operator (of any subtype of `AbstractOperator`) or a 
-        function that returns the Hamiltonian as a function of time.
-- `ψ_0` -- initital state (`Ket`) of a quantum system
-- `t_range` -- time interval for which the system has to be simulated. 
+- `H` -- a function that retrurns the  Hamiltonian operator (of any subtype of `AbstractOperator`) as a function of time.
+- `init_state` -- initital state density matrix (`DenseOperator`) of a quantum system
+- `tspan` -- time interval for which the system has to be simulated. 
         For instance: 
-            t_range=0:10 evaluates the output using time 
-            steps: 0,1,2,...,10. 
-            t_range=0:0.01:1 evaluates the output using 
-            time steps: 0,0.01,0.02,...,1.0 
-
-- `e_ops` -- list of operators for which the expected value 
+            tspan=(0.0,1.0) evaluates the output from t=0.0 to t=1.0
+- `e_ops` -- list of operators (type DenseOperator) for which the expected value 
     (the observables) will be evaluated at each time step in t_range. 
 
-
+- `c_ops` -- list of collapse operators (type DenseOperator). 
 """
-function sesolve(
-        H::AbstractOperator, 
-        ψ_0::Ket{S}, 
-        t_range::StepRangeLen; 
-        e_ops::Vector{T} where {T<:AbstractOperator}=(T)[], 
-    )::Tuple{Vector{Ket{S}},Matrix{<:Real}} where {S<:Complex}
-    Hamiltonian(t)=H
-    sesolve(Hamiltonian, ψ_0, t_range; e_ops=e_ops)
+Base.@kwdef struct LindbladProblem{T<:DenseOperator}
+    H::Function
+    init_state::T
+    tspan::Tuple{Float64,Float64}
+    e_ops::Vector{T}
+    c_ops::Vector{T}
 end
 
+"""
+    sesolve(problem::ShrodingerProblem; kwargs...)
+Solves the Shrodinger equation:
+
+``\\frac{d \\Psi}{d t}=-i \\hat{H}\\Psi``
+
+and returns a tuple correponding the time instance vector, the corresponding wavefunction Ket, and a Vector of observables evaluated at each time step. 
+
+**Fields**
+- `problem` -- An object of type ShrodingerProblem that defines the problem to be solved. 
+- `is_hamiltonian_static` -- A Bool variable indicating whether the Hamiltonian operator changes with time or not. Default value is false. If true, the solver can have significant performance boost.
+- `kwargs` -- list of keyword arguments to be passed to the ODE solver. See (https://docs.sciml.ai/DiffEqDocs/stable/basics/common_solver_opts/#solver_options).
+"""
 function sesolve(
-        H::Function, 
-        ψ_0::Ket{S}, 
-        t_range::StepRangeLen; 
-        e_ops::Vector{T} where {T<:AbstractOperator}=(T)[], 
-    )::Tuple{Vector{Ket{S}},Matrix{<:Real}} where {S<:Complex}
-    dpsi_dt(t,ψ) = -im*H(t)*ψ
-    y=rungekutta2(dpsi_dt, ψ_0, t_range)
-    n_t = length(t_range)
-    n_o = length(e_ops)
-    observable=zeros(n_t, n_o) 
-    for iob in 1:n_o
-        for i_t in 1:n_t
-            observable[i_t, iob] = real(expected_value(e_ops[iob], y[i_t]))
+    problem::ShrodingerProblem;
+    is_hamiltonian_static::Bool = false,
```

Comment:
Wouldn't it be a lot cleaner if we added a specialization to `sesolve`, which takes a static `H`? Rather than passing a flag and doing both cases using a single function doing more than one thing?

Reply:
I agree that duplication is bad, but I don't see why we can't do it cleanly without duplication. Perhaps we could sit together and look at this.

Reply:
As discussed on Teams, we can solve this in a future PR (https://github.com/anyonlabs/Snowflake.jl/issues/242)